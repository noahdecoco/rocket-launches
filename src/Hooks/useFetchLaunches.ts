import { useEffect, useState } from "react";
import axios from "axios";

const API_URL = "https://launchlibrary.net/1.3/launch";

export interface LaunchResponseData {
  error?: string;
  launches?: any[];
}

export interface Dates {
  from: string;
  to: string;
}

const getFormattedDate = (date: Date): string => {
  return date.toISOString().split("T")[0];
};

export const useFetchLaunches = (fromDate?: string, toDate?: string) => {
  const [
    launchesResponseData,
    setLaunchesResponseData,
  ] = useState<null | LaunchResponseData>(null);

  const [dates, setDates] = useState<null | Dates>(null);

  const fetchLaunches = async (url: string) => {
    try {
      const response = await axios.get(url);
      setLaunchesResponseData({
        launches: response.data.launches,
      });
    } catch (error) {
      setLaunchesResponseData({ error: "An error occurred." });
    }
  };

  useEffect(() => {
    // default is to return data for the next three months
    const from = new Date();
    const to = new Date();
    to.setDate(to.getDate() + 90);
    setDates({ from: getFormattedDate(from), to: getFormattedDate(to) });
  }, [fromDate, toDate]);

  useEffect(() => {
    if (dates) {
      fetchLaunches(`${API_URL}/${dates.from}/${dates.to}`);
    }
  }, [dates]);

  return launchesResponseData;
};
