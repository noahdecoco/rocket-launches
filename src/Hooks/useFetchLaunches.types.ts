export interface LaunchResponseData {
  error?: string;
  launches?: MappedLaunchData[];
}

export interface DateRange {
  from: string;
  to: string;
}

export interface PadData {
  id: string;
  name: string;
  longitude: number;
  latitude: number;
}

export interface LaunchData {
  location: {
    pads: PadData[];
  };
}

export interface MappedLaunchData {
  id: string;
  name: string;
  coordinates: [number, number];
}
