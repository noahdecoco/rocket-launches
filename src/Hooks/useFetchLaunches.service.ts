import { LaunchData, MappedLaunchData } from "./useFetchLaunches.types";

export class UseFetchLaunchesService {
  static getFormattedDate(date: Date): string {
    return date.toISOString().split("T")[0];
  }

  static mapResponseForClient(launches: LaunchData[] = []): MappedLaunchData[] {
    try {
      return launches.map(
        (launch): MappedLaunchData => {
          const firstPad = launch.location.pads[0];
          return {
            id: firstPad.id,
            name: firstPad.name,
            coordinates: [firstPad.longitude, firstPad.latitude],
          };
        }
      );
    } catch (error) {
      throw new Error("Oops... Something went wrong!");
    }
  }
}
