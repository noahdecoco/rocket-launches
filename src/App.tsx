import React from "react";

import styles from "./App.module.css";
import { LaunchResponseData, useFetchLaunches } from "./Hooks/useFetchLaunches";

function App() {
  const launches: LaunchResponseData | null = useFetchLaunches();

  return (
    <div className="app">
      <header className={styles.appHeader}>
        <h1 className={styles.appHeaderTitle}>ðŸš€ ðŸ“† Moonshot Calendar Inc.</h1>
      </header>
      <main>
        {!launches ? <div>Loading...</div> : null}

        {launches && launches.error ? <div>{launches.error}</div> : null}

        {launches ? <div>{JSON.stringify(launches, null, 2)}</div> : null}
      </main>
    </div>
  );
}

export default App;
