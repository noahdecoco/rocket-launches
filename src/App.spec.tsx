import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";
import * as hooks from "./Hooks/useFetchLaunches";

describe("<App/>", () => {
  it("renders the title", () => {
    render(<App />);
    const titleElement = screen.getByTestId("app-title");
    expect(titleElement).toBeInTheDocument();
  });

  it("renders the loader only when data is null", () => {
    const wrapper = render(<App />);

    const loaderElement = screen.queryByTestId("app-loader");
    const errorElement = screen.queryByTestId("error-alert");
    const mapContainer = screen.queryByTestId("map-container");

    expect(loaderElement).toBeInTheDocument();
    expect(errorElement).toBeNull();
    expect(mapContainer).toBeNull();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("should render the map when data loads", () => {
    jest.spyOn(hooks, "useFetchLaunches").mockImplementationOnce(() => {
      return {
        launches: [
          {
            name: "test-name",
            id: "1234",
            coordinates: [1234, 4321],
          },
        ],
      };
    });

    const wrapper = render(<App />);

    const loaderElement = screen.queryByTestId("app-loader");
    const errorElement = screen.queryByTestId("error-alert");
    const mapContainer = screen.queryByTestId("map-container");

    expect(loaderElement).not.toBeInTheDocument();
    expect(errorElement).toBeNull();
    expect(mapContainer).toBeInTheDocument();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("should render the empty map and error when data loads", () => {
    jest.spyOn(hooks, "useFetchLaunches").mockImplementationOnce(() => {
      return {
        error: "test-error-message",
      };
    });

    const wrapper = render(<App />);

    const loaderElement = screen.queryByTestId("app-loader");
    const errorElement = screen.queryByTestId("error-alert");
    const mapContainer = screen.queryByTestId("map-container");

    expect(loaderElement).not.toBeInTheDocument();
    expect(errorElement).toBeInTheDocument();
    expect(mapContainer).toBeInTheDocument();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
