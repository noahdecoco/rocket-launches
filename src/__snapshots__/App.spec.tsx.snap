// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders the loader only when data is null 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <header
      class="appHeader"
    >
      <h1
        class="appHeaderTitle"
        data-testid="app-title"
      >
        ðŸš€ ðŸ“† Moonshot Calendar Inc.
      </h1>
    </header>
    <main>
      <div
        class="screen"
        data-testid="app-loader"
      >
        <span
          class="loadingText"
        >
          Loading...
        </span>
      </div>
    </main>
  </div>
</DocumentFragment>
`;

exports[`<App/> should render the empty map and error when data loads 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <header
      class="appHeader"
    >
      <h1
        class="appHeaderTitle"
        data-testid="app-title"
      >
        ðŸš€ ðŸ“† Moonshot Calendar Inc.
      </h1>
    </header>
    <main>
      <div
        class="screen"
        data-testid="error-alert"
      >
        <span
          class="errorText"
        >
          test-error-message
        </span>
      </div>
      <div
        class="mapContainer"
        data-testid="map-container"
      >
        <svg
          class="rsm-svg "
          viewBox="0 0 800 600"
        >
          <g>
            <rect
              fill="transparent"
              height="600"
              width="800"
            />
            <g
              class="rsm-zoomable-group undefined"
              transform="translate(0 0) scale(1)"
            >
              <g
                class="rsm-geographies "
              />
            </g>
          </g>
        </svg>
      </div>
    </main>
  </div>
</DocumentFragment>
`;

exports[`<App/> should render the map when data loads 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <header
      class="appHeader"
    >
      <h1
        class="appHeaderTitle"
        data-testid="app-title"
      >
        ðŸš€ ðŸ“† Moonshot Calendar Inc.
      </h1>
    </header>
    <main>
      <div
        class="mapContainer"
        data-testid="map-container"
      >
        <svg
          class="rsm-svg "
          viewBox="0 0 800 600"
        >
          <g>
            <rect
              fill="transparent"
              height="600"
              width="800"
            />
            <g
              class="rsm-zoomable-group undefined"
              transform="translate(0 0) scale(1)"
            >
              <g
                class="rsm-geographies "
              />
              <g
                class="rsm-marker "
                transform="translate(810.210030067051, 296.41121365738474)"
              >
                <g
                  fill="none"
                  stroke="#FF5533"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  transform="translate(-12, -24)"
                >
                  <circle
                    cx="12"
                    cy="10"
                    r="3"
                  />
                  <path
                    d="M12 21.7C17.3 17 20 13 20 10a8 8 0 1 0-16 0c0 3 2.7 6.9 8 11.7z"
                  />
                </g>
                <text
                  style="font-family: system-ui; fill: #5D5A6D; font-size: 0.6rem;"
                  text-anchor="middle"
                >
                  test-name
                </text>
              </g>
            </g>
          </g>
        </svg>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
